File: Nebokrai/server/src/login/authhandler.cpp
#include "authhandler.h"

#include <utility>

namespace Login
{
namespace
{
constexpr std::int32_t kAuthFailedMessageType = 0x000AF501;
}

AuthHandler::AuthHandler(IAuthHandlerContext& context) noexcept
    : m_Context(context)
{
}

void AuthHandler::OnQuest(const AuthQuest& quest)
{
    static_cast<void>(quest);
    // Точный виртуальный слот 0x004664B0: ret 4.
}

void AuthHandler::OnResponse(const AuthResult& result)
{
    if (result.result == 0) {
        m_Context.PushBackPwdChecked(TagPwdChecked(
            result.clientSocketId,
            result.clientIp,
            result.account,
            {},
            false));
        return;
    }

    const std::int8_t responseCode = ResponseCode(result.result);
    LoginNet::CMessage response(kAuthFailedMessageType);
    response.Base().Add(static_cast<char>(responseCode));
    auto sendResult = m_Context.SendToClient(response, result.clientSocketId);
    if (const auto* error = std::get_if<LoginNet::SendMessageError>(&sendResult)) {
        m_Context.PushAuthHandlerNotice(AuthHandlerNotice{
            .kind = AuthHandlerNoticeKind::ClientResponseFailed,
            .socketId = result.clientSocketId,
            .account = {},
            .sendError = *error,
        });
    }

    if (responseCode != 7) {
        return;
    }

    const PasswordFailureOutcome failure =
        m_Context.RegisterPasswordFailure(result.account);
    if (failure.kind == PasswordFailureOutcomeKind::BanAttempted &&
        !failure.banSucceeded) {
        m_Context.PushAuthHandlerNotice(AuthHandlerNotice{
            .kind = AuthHandlerNoticeKind::CdKeyBanFailed,
            .account = result.account,
        });
    } else if (failure.kind == PasswordFailureOutcomeKind::BanOwnerMissing) {
        m_Context.PushAuthHandlerNotice(AuthHandlerNotice{
            .kind = AuthHandlerNoticeKind::CdKeyBanOwnerMissing,
            .account = result.account,
        });
    }
}

std::int8_t AuthHandler::ResponseCode(std::int32_t result) noexcept
{
    switch (result) {
    case 2: return 5;
    case 3: return 6;
    case 4: return 73; // 'I'
    case 5: return 18;
    case 6: return 63; // '?'
    case 8: return 82; // 'R'
    case 9: return 83; // 'S'
    case 10: return 84; // 'T'
    case 11: return 87; // 'W'
    case 12: return 88; // 'X'
    default: return 7;
    }
}
}

File: Nebokrai/server/rust/src/loginserver/loginserver/authhandler.rs
//! Владелец ответа AuthServer `AuthHandler` из `authhandler.cpp`.
//!
//! Контракт обеих virtual-функций подтверждён точной парой LoginServer EXE/PDB.
//! Первый slot, вызываемый `AuthManager::addQuest`, является no-op. `OnResponse` при
//! результате `0` создаёт owned `TagPwdChecked` с пустым world-name и
//! `has_matrix = false`, затем передаёт его в `CLoginQueue`. При совпадении
//! account очередь вызывает `CGame::KickOut` под исходным queue-lock, удаляет
//! прежний объект и добавляет новый в хвост.
//!
//! Любой ненулевой результат сначала отправляет клиенту `0xAF501` с одним
//! байтом. Коды `2..=6` и `8..=12` имеют отдельное отображение;
//! `1`, `7` и остальные значения идут через исходный default `7`. Только
//! default-ветвь после client-send изменяет счётчик неверного пароля. Первая
//! ошибка записывает `1`, последующие увеличивают значение, пока старое
//! значение меньше лимита; следующая ошибка вызывает `CRsCDKey::CDKeyBan` и
//! удаляет счётчик независимо от результата вызова.
//!
//! В оригинале `mAuthHandler` был встроен в `CGame`, а callbacks находили тот
//! же глобальный `CGame`. Самоссылочная Rust-структура для этого не создаётся:
//! `CGame` реализует технический `AuthListener`-адаптер и немедленно передаёт
//! оба вызова этому stateless owner. Такая форма API меняет только владение,
//! но сохраняет синхронный порядок callbacks. `std::string`, `std::map`, SEH,
//! ручные `new/delete` и compiler cleanup удалены как технический шум; их
//! существенные эффекты выражены owned bytes, владельцами и `Drop`.

use crate::loginserver::loginserver::authmanager::{AuthQuest, AuthResult};
use crate::loginserver::loginserver::game::{AuthHandlerNotice, CGame, PasswordFailureOutcome};
use crate::loginserver::loginserver::loginqueue::TagPwdChecked;
use crate::nets::netlogin::message::CMessage;

const AUTH_FAILED_MESSAGE_TYPE: i32 = 0x000A_F501;

/// Stateless-владелец двух virtual slot исходного embedded `AuthHandler`.
pub(crate) struct AuthHandler;

impl AuthHandler {
    /// Сохраняет доказанный no-op первого virtual slot.
    pub(crate) fn on_quest(_quest: &AuthQuest) {}

    /// Выполняет полный доказанный switch исходного `OnResponse`.
    pub(crate) fn on_response(game: &mut CGame, result: &AuthResult) {
        if result.result == 0 {
            let checked = TagPwdChecked::new(
                result.client_socket_id,
                result.client_ip,
                result.account.clone(),
                Vec::new(),
                false,
            );
            game.push_back_pwd_checked(checked);
            return;
        }

        let response_code = response_code(result.result);
        let mut message = CMessage::new(AUTH_FAILED_MESSAGE_TYPE);
        message.base_mut().add_char(response_code);
        if let Err(error) = game.send_to_client(&message, result.client_socket_id) {
            game.push_auth_handler_notice(AuthHandlerNotice::ClientResponseFailed {
                socket_id: result.client_socket_id,
                error,
            });
        }

        if response_code != 7 {
            return;
        }

        match game.register_password_failure(&result.account) {
            PasswordFailureOutcome::BanAttempted { succeeded: false } => {
                game.push_auth_handler_notice(AuthHandlerNotice::CdKeyBanFailed {
                    account: result.account.clone(),
                });
            }
            PasswordFailureOutcome::BanOwnerMissing => {
                game.push_auth_handler_notice(AuthHandlerNotice::CdKeyBanOwnerMissing {
                    account: result.account.clone(),
                });
            }
            PasswordFailureOutcome::Disabled
            | PasswordFailureOutcome::Counted { .. }
            | PasswordFailureOutcome::BanAttempted { succeeded: true } => {}
        }
    }
}

const fn response_code(result: i32) -> i8 {
    match result {
        2 => 5,
        3 => 6,
        4 => 73,
        5 => 18,
        6 => 63,
        8 => 82,
        9 => 83,
        10 => 84,
        11 => 87,
        12 => 88,
        _ => 7,
    }
}
